# Exercise 03

## Data preparation and model evaluation exercise with credit scoring

Banks play a crucial role in market economies. They decide who can get finance and on what terms and can make or break investment decisions. For markets and society to function, individuals and companies need access to credit. 

Credit scoring algorithms, which make a guess at the probability of default, are the method banks use to determine whether or not a loan should be granted. This competition requires participants to improve on the state of the art in credit scoring, by predicting the probability that somebody will experience financial distress in the next two years. [Dataset](https://www.kaggle.com/c/GiveMeSomeCredit)

Attribute Information:

|Variable Name	|	Description	|	Type|
|----|----|----|
|SeriousDlqin2yrs	|	Person experienced 90 days past due delinquency or worse 	|	Y/N|
|RevolvingUtilizationOfUnsecuredLines	|	Total balance on credit divided by the sum of credit limits	|	percentage|
|age	|	Age of borrower in years	|	integer|
|NumberOfTime30-59DaysPastDueNotWorse	|	Number of times borrower has been 30-59 days past due |	integer|
|DebtRatio	|	Monthly debt payments	|	percentage|
|MonthlyIncome	|	Monthly income	|	real|
|NumberOfOpenCreditLinesAndLoans	|	Number of Open loans |	integer|
|NumberOfTimes90DaysLate	|	Number of times borrower has been 90 days or more past due.	|	integer|
|NumberRealEstateLoansOrLines	|	Number of mortgage and real estate loans	|	integer|
|NumberOfTime60-89DaysPastDueNotWorse	|	Number of times borrower has been 60-89 days past due |integer|
|NumberOfDependents	|	Number of dependents in family	|	integer|


Read the data into Pandas

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
with zipfile.ZipFile('../datasets/KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.io.parsers.read_table(f, sep=',')

data.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                 1                              0.766127  45.0   
1           1                 0                              0.957151  40.0   
2           2                 0                              0.658180  38.0   
3           3                 0                              0.233810  30.0   
4           4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [2]:
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)
data['SeriousDlqin2yrs'].value_counts()

0    105299
1      7616
Name: SeriousDlqin2yrs, dtype: int64

In [3]:
X

Unnamed: 0  RevolvingUtilizationOfUnsecuredLines   age  \
0                0                              0.766127  45.0   
1                1                              0.957151  40.0   
2                2                              0.658180  38.0   
3                3                              0.233810  30.0   
4                4                              0.907239  49.0   
5                5                              0.213179  74.0   
6                6                              0.754464  39.0   
7                7                              0.189169  57.0   
8                8                              0.644226  30.0   
9                9                              0.018798  51.0   
10              10                              0.010352  46.0   
11              11                              0.964673  40.0   
12              12                              0.548458  64.0   
13              13                              0.166284  53.0   
14              14                              0.221813  43.0   
15              15                              0.602794  25.0   
16              16                              0.200923  43.0   
17              17                              0.025656  38.0   
18              18                              1.000000  39.0   
19              19                              0.075427  32.0   
20              20                              0.046560  58.0   
21              21                              0.052436  58.0   
22              22                              0.034421  69.0   
23              23                              0.452516  24.0   
24              24                              0.392995  58.0   
25              25                              0.704074  28.0   
26              26                              1.000000  24.0   
27              27                              0.186869  57.0   
28              28                              0.693330  42.0   
29              29                              0.000000  64.0   
...            ...                                   ...   ...   
112885      112885                              0.920107  31.0   
112886      112886                              1.005733  48.0   
112887      112887                              1.010934  63.0   
112888      112888                              0.081930  57.0   
112889      112889                              0.983825  55.0   
112890      112890                              0.075535  43.0   
112891      112891                              0.025449  58.0   
112892      112892                              0.058001  83.0   
112893      112893                              0.071273   NaN   
112894      112894                              1.026395  44.0   
112895      112895                              0.962721  61.0   
112896      112896                              0.917635  52.0   
112897      112897                              0.224711  55.0   
112898      112898                              0.067644  64.0   
112899      112899                              0.810012  43.0   
112900      112900                              0.021046  37.0   
112901      112901                              0.002485  82.0   
112902      112902                              0.954409  26.0   
112903      112903                              0.168102  49.0   
112904      112904                              1.000000  28.0   
112905      112905                              0.902051  31.0   
112906      112906                              0.013356  62.0   
112907      112907                              0.055518  46.0   
112908      112908                              0.104112  59.0   
112909      112909                              1.000000  22.0   
112910      112910                              0.385742  50.0   
112911      112911                              0.040674  74.0   
112912      112912                              0.299745  44.0   
112913      112913                              

# Exercise 3.1

Input the missing values of the Age and Number of Dependents 

Es importante antes de correr algun modelo o hacer algun tratamiento a la data, revisar los valores fatantes. a continuacion se observa que la variable age,y NumberOfDependents contiene valores missing. con 4267 registros en un total de 112.915 registros, es decir un 3,77% aproximadamente. por lo cual se llenan dichos valores con la mediana para no perder dichos registros. se toma la mediana o percentil 50  porque es una medida que nos indica la dispersión de los datos. por otro lado, se evidencia que para age existe un caso con age=0 como valor minimo el cual deberia excluirse y como edad maxima 103.

In [4]:
# check for missing values
data.isnull().sum()

Unnamed: 0                                 0
SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

In [5]:
data.shape

(112915, 12)

In [6]:
data['age'].describe()

count    108648.000000
mean         51.361304
std          14.453571
min           0.000000
25%          41.000000
50%          51.000000
75%          62.000000
max         103.000000
Name: age, dtype: float64

In [7]:
#data['age'].value_counts()

In [8]:
# fill missing values for age with the median age and number of dependents
data.age.fillna(data.age.median(), inplace=True)
data.NumberOfDependents.fillna(data.NumberOfDependents.median(), inplace=True)

In [9]:
# check for missing values
data.isnull().sum()

Unnamed: 0                              0
SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

In [10]:
data.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                 1                              0.766127  45.0   
1           1                 0                              0.957151  40.0   
2           2                 0                              0.658180  38.0   
3           3                 0                              0.233810  30.0   
4           4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [11]:
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)

In [12]:
data['SeriousDlqin2yrs'].value_counts()

0    105299
1      7616
Name: SeriousDlqin2yrs, dtype: int64

# Exercise 3.2

From the set of features

Select the features that maximize the **F1Score** the model using K-Fold cross-validation


Para el calculo y la intepretacion se debe tener en cuenta las siguientes definiciones:

Condition positive (P) : the number of real positive cases in the data Condition negative (N) : the number of real negative cases in the data True positive (TP) True negative (TN) False positive (FP): Type I error False negative (FN) : eqv. with miss, Type II error

accuracy (ACC) = (TP + TN)/ P + N = (TP + TN)/ TP + TN + FP + FN

Definicion: F1= 2 (PPV TPR / PPV + TPR) = 2PT / (2TP + FP + FN)

Se cargan los paquetes que se van a utilizar y se crean los folds. Adicionalmente, se parte la base en train y test para después poder correr la regresión logística con las distintas bases 10 veces cada set compuesto de 25 observaciones. 

In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [14]:
# se procece a realizar cross validation 

# simulate splitting a dataset of 25 observations into 5 folds
from sklearn.cross_validation import KFold
kf = KFold(25, n_folds=5, shuffle=False)

# print the contents of each training and testing set
print('{} {:^61} {}'.format('Iteration', 'Training set observations', 'Testing set observations'))
for iteration, data in enumerate(kf, start=1):
    print('{:^9} {} {:^25}'.format(str(iteration), str(data[0]), str(data[1])))

Iteration                   Training set observations                   Testing set observations
    1     [ 5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]        [0 1 2 3 4]       
    2     [ 0  1  2  3  4 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]        [5 6 7 8 9]       
    3     [ 0  1  2  3  4  5  6  7  8  9 15 16 17 18 19 20 21 22 23 24]     [10 11 12 13 14]     
    4     [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 20 21 22 23 24]     [15 16 17 18 19]     
    5     [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]     [20 21 22 23 24]     


C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [15]:
# Create k-folds
kf = KFold(X.shape[0], n_folds=10, random_state=0)

results = []

for train_index, test_index in kf:
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]


In [16]:
 # train a logistic regression model
logreg = LogisticRegression(C=1e9)
logreg.fit(X_train, y_train)

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=None,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

In [17]:
 # make predictions for testing set
y_pred_class = logreg.predict(X_test)

In [18]:
 # calculate testing accuracy
results.append(metrics.accuracy_score(y_test, y_pred_class))
print(metrics.accuracy_score(y_test, y_pred_class))

#ACC= (10483 + 9)/ TOTAL DE DATOS = 0,92

0.9292356744309627


In [19]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred_class)

array([[10483,     6],
       [  793,     9]], dtype=int64)

In [20]:
from sklearn.cross_validation import cross_val_score

logreg = LogisticRegression(C=1e9)


In [21]:
# en todas ls iteracciones no varian mucho los resultados frente a las otras, el ACC esta alrededor del 93%

results = cross_val_score(logreg, X, y, cv=10, scoring='accuracy')
results

array([0.93260716, 0.93269571, 0.93234148, 0.93269571, 0.93269571,
       0.93243004, 0.93295545, 0.93295545, 0.93330972, 0.93286094])

In [22]:
print(results.mean())

0.9327547373582709


In [23]:
from sklearn.metrics import precision_score, recall_score, f1_score
print('precision_score ', precision_score(y_test, y_pred_class))
print('recall_score    ', recall_score(y_test, y_pred_class))
print('f1_score    ', f1_score(y_test, y_pred_class))

precision_score  0.6
recall_score     0.011221945137157107
f1_score     0.022031823745410035


Conclusiones generales del primer ejercicio: aunque el ACC esta alrededor del 93%  promedio en todas las muestras usando K=10 con cross validation, aparentemente es un buen modelo. Sin embargo el f1_score es tan solo del 2%, lo que siginifica que la clase negativa no esta siendo identificada por el modelo. Lo anterior,  debido a que la clase o variable SeriousDlqin2yrs  esta desbalanceada con el 6,76%.

# Exercise 3.3

Now which is the best set of features selected by AUC

In [24]:
import numpy as np
import pandas as pd
np.set_printoptions(suppress=True)
pt = np.get_printoptions()['threshold']
from sklearn.feature_selection import VarianceThreshold

In [25]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
with zipfile.ZipFile('../datasets/KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.io.parsers.read_table(f, sep=',')

data.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                 1                              0.766127  45.0   
1           1                 0                              0.957151  40.0   
2           2                 0                              0.658180  38.0   
3           3                 0                              0.233810  30.0   
4           4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [26]:
sel = VarianceThreshold(threshold=(.8 * (1 - .8)))
feature_set = sel.fit_transform(X)

In [27]:
X[0:5]

Unnamed: 0  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                              0.766127  45.0   
1           1                              0.957151  40.0   
2           2                              0.658180  38.0   
3           3                              0.233810  30.0   
4           4                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [28]:
data[0:5]

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                 1                              0.766127  45.0   
1           1                 0                              0.957151  40.0   
2           2                 0                              0.658180  38.0   
3           3                 0                              0.233810  30.0   
4           4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [29]:
for i in range(2,len(data.columns)):
    X=data.iloc[:,i:]
    y=data.SeriousDlqin2yrs
    logreg=LogisticRegression(C=1e9)
    results=cross_val_score(logreg,X, y, cv=3, scoring='roc_auc')
    print("Variable %s-> ROC_AUC: %s" % (len(data.columns)-i, pd.Series(results).mean()))

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

Conclusion:

Como se puede observar, el modelo que ofrece el mejor AUC es el que contiene 9 variables, pues es el que contabiliza el 69%.